File: esc180/fall semester/assignments/lab9/ge.c
#include<stdlib.h>
#include<stdio.h>
int ge_fw(float *matrix, int rows, int cols, float *matrix_out);
int ge_bw(float *matrix, int rows, int cols, float *matrix_out);


int ge_fw(float *matrix, int rows, int cols, float *matrix_out){
	int r = 0;
	int c = 0;
	int i = 0;
	float temp = 0.0;
	int x = 0;
	int k = 0;
	int j = 0;
	float factor = 0.0;

	if(matrix == NULL){
		return -1;
	}
	if(matrix_out == NULL){
		return -1;
	}
	
	/*copying matrix into matrix_out*/
	for(r = 0; r < rows; r++){
		for(c = 0; c < cols; c++){
			matrix_out[r*cols + c] = matrix[r*cols + c];
		}
	}

	/*find first row with a non-zero entry and exchange with first row*/
	for(r = 0;r < rows;r++){
		if(matrix_out[r*cols] != 0.0){
			if(r*cols != 0){
				/*swap rows*/
				for(i = 0; i < cols; i++){
					temp = matrix_out[r*cols + i];
					matrix_out[r*cols + i] = matrix_out[i];
					matrix_out[i] = temp;				
				}
			}
			break;
		}
	}

	for(x = 0; x < cols; x++){
		for(r = k + 1; r < rows; r++){
			/*if another nonzero beginning row is found*/	
			if(matrix_out[r*cols + x]  == 0){
				for(j = r; j < rows; j++){
					if(matrix_out[j*cols] != 0){
						/*swap rows*/
						for(i = 0; i < cols; i++){
							temp = matrix_out[j*cols + i];
							matrix_out[j*cols + i] = matrix_out[r*cols + i];
							matrix_out[r*cols + i] = temp;
						}
						break;
					}
				}
			}

			factor = matrix_out[r*cols + x]/matrix_out[k*cols + x];

			/*apply factor to the columns of that row*/

			for(c = x; c < cols; c++){
				matrix_out[r*cols + c] -= (factor * matrix_out[k*cols + c]);
			}
		}
		k++;
	}

	return 0;			
							
}

int ge_bw(float *matrix, int rows, int cols, float *matrix_out){
	int r = 0;
	int c = 0;
	int x = 0;
	int lastNonZeroRow = 0;
	float factor = 0.0;
	int factorFound = 0;
	int leadingOneCol = 0;

	/*matrix_out is already changed according to ge_fw matrix is the original matrix*/

	if(matrix_out == NULL){
		return -1;
	}
	if(matrix == NULL){
		return -1;
	}

	/*starting with the last nonzero row, normalize the row by its first nonzero entry*/
	
	/*find the last nonzer row*/
	lastNonZeroRow = -1;
	for(r = rows - 1; r >= 0; r--){
		for(c = 0; c < cols; c++){
			if(matrix_out[r*cols +c] != 0.0){
				lastNonZeroRow = r;
				break;
			}
		}
		if(lastNonZeroRow > -1){
			break;
		}
	}

	/*you want the first nonzero column in the row to be one*/
	for(r = lastNonZeroRow; r >= 0; r--){
		factor = 1;
		factorFound = 0;
		for(c = 0; c < cols; c++){
			if(matrix_out[r*cols + c] != 0.0 && factorFound == 0){
				factor = matrix_out[r*cols + c];
				factorFound = 1;
			}
			matrix_out[r*cols + c] = matrix_out[r*cols + c]/factor;
		}
	}

	/*add multiples of this row to previous rows so that all the entries in the column
 	 *containing leading 1s are zero*/

	for(r = lastNonZeroRow; r > 0; r--){
		/*find column that contains leading 1*/
		for(c = 0; c < cols; c++){
			if(matrix_out[r*cols + c] == 1.0){
				leadingOneCol = c;
				break;
			}
		}
	
		for(c = r-1; c >= 0; c--){
			factor = matrix_out[c*cols + leadingOneCol];
			for(x = leadingOneCol; x < cols; x++){
				matrix_out[c*cols + x] -= factor * matrix_out[r*cols + x];
			}
		}
	}		 	
	return 0;
}

File: esc180/fall semester/assignments/lab9/ge.py
def ge_fw(matrix):
	#find first row with a non-zero entry and exchange with first row
	m = list(matrix)
	
	if len(m) < 1:
		return m

	for i in range(len(m)):
		if m[i][0] != 0:
			temp = m[i]
			m[i] = m[0]
			m[0] = temp
			break

	#add multiples of the first row to lower rows so that the lower entries ofthe first column are all zero
	#form the submatrix arising from deleting the first col and row and applying the forward step

	n_cols = len(m[0])
	n_rows = len(m)
	k = 0

	for x in range(n_cols):	
		for i in range(k + 1, len(m)):	#loop through rows			
			if m[i][x] == 0:
				for z in range(i, len(m)):
					if m[z][0] != 0:
						temp = m[z]
						m[z] = m[i]
						m[i] = temp
						break	
			
			factor = m[i][x]/m[k][x]

			#apply factor to columns of that row
			for j in range(x,len(m[i])):
				m[i][j] -= factor * m[k][j]
		k += 1
	return m

def ge_bw(matrix):
	m = list(matrix)
	
	if len(m) < 1:
		return m

	#starting with the last nonzero row, normalize the row by its first nonzero entry

	
	#find the last non-zero row
	lastNonZeroRow = -1

	for i in range(len(m)-1, -1, -1):
		for j in range(len(m[i])):
			if m[i][j] != 0:
				lastNonZeroRow = i
		if lastNonZeroRow > -1:
			break



	#you want the first nonzero column in the row to be one (multiply row by a nonzero factor)
	for j in range(lastNonZeroRow, -1, -1):		
		factor = 1
		factorFound = False
		for i in range(len(m[j])):
			if m[j][i] != 0 and not factorFound:
				factor = m[j][i]
				factorFound = True

			m[j][i] = m[j][i]/factor  


	#add multiples of this row to previous rows so that all the entries in the column containing leading 1s are zero
	for i in range(lastNonZeroRow, 0, -1):

		#find column that contains leading 1
		leadingOneCol = 0
	
		for j in range(len(m[i])):
			if m[i][j] == 1:
				leadingOneCol = j
				break


		for j in range(i-1, -1, -1):
			factor = m[j][leadingOneCol];
			for x in range(leadingOneCol, len(m[i]), 1):				
				m[j][x] -= factor * m[i][x]
				
	return m

